core/vdbe/execute.rs

```diff
@@ -1316,42 +1378,200 @@ pub fn op_column(
                     must_be_btree_cursor!(*cursor_id, program.cursor_ref, state, "Column");
                 let cursor = cursor.as_btree_mut();
-                let record = return_if_io!(cursor.record());
 
-                let Some(record) = record.as_ref() else {
+                if cursor.get_null_flag() {
                     break 'value Value::Null;
+                }
+
+                let record_result = return_if_io!(cursor.record());
+                let Some(record) = record_result.as_ref() else {
+                    break 'value default.clone().unwrap_or(Value::Null);
                 };
 
-                if cursor.get_null_flag() {
-                    break 'value Value::Null;
+                let payload = record.get_payload();
+
+                if payload.is_empty() {
+                    break 'value default.clone().unwrap_or(Value::Null);
+                }
+
+                let mut record_cursor = cursor.record_cursor.borrow_mut();
+
+                if record_cursor.serial_types.is_empty() && record_cursor.offsets.is_empty() {
+                    let (header_size, header_len_bytes) = read_varint_fast(payload)?;
+                    let header_size = header_size as usize;
+
+                    if header_size > payload.len() || header_size > 98307 {
+                        return Err(LimboError::Corrupt("Header size exceeds bounds".into()));
+                    }
+
+                    record_cursor.header_size = header_size;
+                    record_cursor.header_offset = header_len_bytes;
+
+                    record_cursor.offsets.push(header_size);
                 }
 
-                if let Some(value) = record.get_value_opt(*column) {
-                    break 'value value.to_owned();
+                let target_column = *column;
+                let mut parse_pos = record_cursor.header_offset;
+                let mut data_offset = record_cursor
+                    .offsets
+                    .last()
+                    .copied()
+                    .unwrap_or(record_cursor.header_size);
+
+                // Adjust data_offset if we already have some columns parsed
+                if !record_cursor.serial_types.is_empty() && !record_cursor.offsets.is_empty() {
+                    data_offset = *record_cursor.offsets.last().unwrap();
                 }
 
-                default.clone().unwrap_or(Value::Null)
+                // Parse the header for serial types incrementally until we have the target column
+                while record_cursor.serial_types.len() <= target_column
+                    && parse_pos < record_cursor.header_size
+                    && parse_pos < payload.len()
+                {
+                    let (serial_type, varint_len) = read_varint_fast(&payload[parse_pos..])?;
+
+                    record_cursor.serial_types.push(serial_type);
+                    parse_pos += varint_len;
+                    let data_size = match serial_type {
```

Comment:
can we use the `SerialType` newtype wrapper for this? we have this same logic elsewhere